R/featEng-tails.R

```diff
@@ -0,0 +1,166 @@
+#'
+#' @details \code{computeTailMeans()} finds the mean of the alpha-percentile tails
+#' of each individual patient through the following steps: first, the function
+#' extracts the 'PID,', 'x', 'y', 'z', 'value,' and 'cluster' columns from
+#' \code{voxel_df} and \code{data_df}. We run \code{cbind} to combine these specific
+#' columns. Following this, we isolate each individual cluster label and then
+#' extract each individual PID. With each individual patient, we then find the
+#' upper and lower alpha-percentile cutoffs, before finding the means of the
+#' values that lie above and below those cutoffs, respectively. We then collect
+#' these results and put them into a dataframe. If \code{data_df} is not specified,
+#' then the function will assume that all of the data comes from a single cluster.
+#'
+#' @param voxel_df A \code{dataframe} in the following format: the first column
+#' is assumed to be an index column counting from 1 to \code{nrow(voxel_df)},
+#' and the second column is a column labeled "PID," which contains patient
+#' identification. The third, fourth, and fifth columns are 'x', 'y', and 'z'
+#' columns. The sixth column contains the value of the voxel that corresponds
+#' with the given x-y-z coordinate.
+#' @param data_df A \code{tibble} that is meant to be outputted from \code{dataDrivenClusters()}.
+#' This tibble should be in the following format: the first three columns are
+#' assumed to be 'x', 'y', and 'z' coordinates. The fourth and fifth columns are
+#' UMAP coordinates, and the fifth column contains cluster labels. If this is
+#' equal to \code{null}, \code{computeTailMeans()} will assume that all data comes
+#' from the same cluster.
+#' @param alpha A \code{double} that indicates what quantiles should be extracted
+#' from the patient data. The lower the number, the tighter the tail will be.
+#'
+#' @export
+
+computeTailMeans <- function(voxel_df, data_df = NULL, alpha = 0.05) {
+
+  # assuming long format voxel_df in which column 1 is index, column 2 is PID,
+  # column 3-5 is x,y,z, and column 6 is value. Assuming data_df to be output
+  # from dataDrivenClusters(), in which columns 1-3 are x,y,z, column 4-5 are
+  # UMAP coords, and column 6 is cluster labels.
+
+  if (alpha != 0 & alpha != 1) {
+
+    if (!is.null(data_df)) {
+
+      # cbind voxel_df and data_df
+      combine_df <- cbind(voxel_df, data_df)
+
+      # extracting the columns in following order: PID, x, y, z, Value, Cluster
+      combind_df <- combine_df[, c(2, 3, 4, 5, 6, 12)]
+
+      # adding additional xyz identifier column
+      combine_df <- combine_df %>% mutate(xyz = paste0(combine_df[, 1],", ",combine_df[, 2],", ",combine_df[, 3]))
+
+      #removing x, y, z columns since those are redundant after the xyz column.
+      combine_df <- combine_df[, c(1, 5, 6, 7)]
+
+      num_clusts <- length(unique(combine_df[, 2]))
+
+      # initializing empty vectors
+      clust_vec <- c()
```

Comment:
These should be vectorized. It is slower to iteratively fill a vector by doing `v <- c(v, x)` over and over again. Instead, you should pre-specify the lengths of the vectors, such as `clust_vec <- rep(NA, num_clusts)`.

The remaining vectors should be of the same length, so you should initialize them as
`PID_vec  <- lower_mean_vec  <- upper_mean_vec <- etc... <- rep(NA, length(blah))` where `blah <- unique(combine_df[, 1])` and should be an informative variable name

Reply:
Should `clust_vec` have the same length as all the other vectors, since we're appending it to a tibble at the end anyways?

Reply:
No, in this case it should just be `i`. R auto-populates it